### PR TITLE
Move CICD/ArgoCD out out Environments folder to "Config"

### PR DESCRIPTION
**Pull request by gaganhegde**
**What type of PR is this?**
 /kind design
carbon

**Which issue(s) this PR fixes**:
https://issues.redhat.com/browse/GITOPS-147

Co Authored by :
@ishitasequeira

### CODE REVIEW

File: pkg/pipelines/argocd/argocd.go
Comment (bigkevmcd) [inline pkg/pipelines/argocd/argocd.go]: This should probably be renamed as it's no longer an `argoEnv` ?

File: pkg/pipelines/argocd/argocd.go
Comment (bigkevmcd) [inline pkg/pipelines/argocd/argocd.go]: This line is not doing anything ?

File: pkg/pipelines/argocd/argocd.go
Comment (bigkevmcd) [inline pkg/pipelines/argocd/argocd.go]: This line can come out too.

File: pkg/pipelines/argocd/argocd.go
Comment (bigkevmcd) [inline pkg/pipelines/argocd/argocd.go]: You're not using `err` here, but you can remove this, as it now returns `nil` if it doesn't exist, and there can't be any more than one now.

File: pkg/pipelines/argocd/argocd.go
Comment (bigkevmcd) [inline pkg/pipelines/argocd/argocd.go]: You can simplify this to `k := range files {`

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config.go]: Couple of things here...

I'd rename `Cicd` to `CICDConfig`, or possibly `PipelinesConfig`.

It doesn't need to return an error, if you look at the code, it never returns a non-nil case, but you don't need to, as that was for when you could have multiple environments.

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config.go]: See above for the CICD case...no error, and maybe rename to `ArgoCDConfig`?

File: pkg/pipelines/config/config_test.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config_test.go]: This can be removed?

File: pkg/pipelines/config/config_test.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config_test.go]: You can now simplify this entirely, we only need one case, `testEnv` can go away.

File: pkg/pipelines/config/config_test.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config_test.go]: These lines can be removed?

File: pkg/pipelines/config/config_test.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config_test.go]: Again, see above.

File: pkg/pipelines/config/config_test.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config_test.go]: Some more lines that can come out?

File: pkg/pipelines/config/config_test.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config_test.go]: And here too...

File: pkg/pipelines/config/config_test.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config_test.go]: We can remove this?

File: pkg/pipelines/config/validate.go
Comment (bigkevmcd) [inline pkg/pipelines/config/validate.go:88]: It looks like you're not using `invalidEnvironment` anywhere now?

File: pkg/pipelines/config/validate.go
Comment (bigkevmcd) [inline pkg/pipelines/config/validate.go]: m.Walk can return an error, and it's not being picked up here.

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config.go]: This doesn't seem to be true anymore?

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config.go]: This comment doesn't seem to be right?

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config.go]: Not sure this one is right either?

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [reply pkg/pipelines/config/config.go]: Also, the `Environment` comment refers to the CICD environments.

**Review by bigkevmcd — CHANGES_REQUESTED** [review]

File: pkg/pipelines/config/config_test.go
Comment (wtam2018) [inline pkg/pipelines/config/config_test.go:79]: Will this test be replaced by a new one?

File: pkg/pipelines/config/validate.go
Comment (wtam2018) [inline pkg/pipelines/config/validate.go]: This check you have here is checking for namespace collision between any plain old Environment and "specail" CICD/ArgoCD which is a good validation.   However, we got here not because we are adding any applications so the error messages are  off.    Also, we should use the error functions like other errors which log the env.Path (otherwise, good luck troubleshooting what went wrong with the manifest file.)

You should also add unit tests for these cases.

Instead of keeping manifest in vw, I would keep the namespaces of "config" (i.e. CICD and ArgoCD namespaces).  Then, your checking code here wouldn't need to reference the config namespaces from manifest.   It will make the code looks cleaner.   It means that you should walk the Config in config.walk() which is another problem.   You need to validate the names of CICD and ArgoCD environments which have been skipped now.

You should add unit tests for these validation.

File: pkg/pipelines/environments/environments.go
Comment (wtam2018) [inline pkg/pipelines/environments/environments.go]: Please delete.

File: pkg/pipelines/argocd/argocd.go
Comment (wtam2018) [inline pkg/pipelines/argocd/argocd.go]: It's argocd.  Or you mean https://en.wikipedia.org/wiki/Argo_(2012_film)

File: pkg/pipelines/config/config.go
Comment (wtam2018) [inline pkg/pipelines/config/config.go:117]: Argo or ArgoCD?

File: pkg/pipelines/config/config.go
Comment (wtam2018) [inline pkg/pipelines/config/config.go]: If CICD and ArgoCD are no longer Environments, we should probably stop calling them "Environments".

File: pkg/pipelines/service_test.go
Comment (wtam2018) [inline pkg/pipelines/service_test.go]: Please clean this up.   It's pretty messy.

File: pkg/pipelines/tekton_builder.go
Comment (wtam2018) [inline pkg/pipelines/tekton_builder.go]: Please delete.

File: pkg/pipelines/config/config.go
Comment (wtam2018) [inline pkg/pipelines/config/config.go]: We should call it "Name" like any entities in the manifest.   We just create namespaces for them but they are names of something (like Env, App).  In this case, it is a Config.

File: pkg/pipelines/config/config.go
Comment (wtam2018) [inline pkg/pipelines/config/config.go:126]: We should call it "Name" like any entities in the manifest.   We just create namespaces for them but they are names of something (like Env, App).  In this case, it is a Config.

File: pkg/pipelines/environments/environments_test.go
Comment (wtam2018) [inline pkg/pipelines/environments/environments_test.go]: The caller downstream does not handle nil.   This will cause the caller to crash.

**Review by wtam2018 — CHANGES_REQUESTED** [review]
I would like to have another look after all issues have been resolved.

File: pkg/pipelines/config/config.go
Comment (gaganhegde) [reply pkg/pipelines/config/config.go]: Changed to PathForArgoCdConfig() and PathForCICDConfig().

File: pkg/pipelines/config/config.go
Comment (gaganhegde) [reply pkg/pipelines/config/config.go]: Now changed to Name.

Comment (gaganhegde) [reply pkg/pipelines/config/config.go]: Now Changed to Name.

File: pkg/pipelines/argocd/argocd.go
Comment (gaganhegde) [reply pkg/pipelines/argocd/argocd.go]: Now changed to ArgoCDConfig and PipelineConfig

File: pkg/pipelines/config/config_test.go
Comment (gaganhegde) [reply pkg/pipelines/config/config_test.go:79]: This wont be necessary as the the cicd environment is now present in a separate  struct, so the sorting is not necessary.

File: pkg/pipelines/argocd/argocd.go
Comment (gaganhegde) [reply pkg/pipelines/argocd/argocd.go]: Now changed to ArgoCDConfig and PipelinesConfig.

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [reply pkg/pipelines/config/config.go]: I'm not so sure about this.

ArgoCD it should at least be the namespace, as it's the namespace that the ArgoCD Applications need to go into (which is the ArgoCD namespace).

The CICD one is perhaps different, but at least for argo, we want folks to realise that it's the _namespace_.

File: pkg/pipelines/bootstrap.go
Comment (wtam2018) [inline pkg/pipelines/bootstrap.go]: Why are we adding this "stage" Environemnt in this PR?

File: pkg/pipelines/config/config.go
Comment (wtam2018) [inline pkg/pipelines/config/config.go]: lower caases json declaration.

File: pkg/pipelines/config/config.go
Comment (wtam2018) [inline pkg/pipelines/config/config.go]: Please delete.

File: pkg/pipelines/config/config.go
Comment (bigkevmcd) [inline pkg/pipelines/config/config.go]: This could be tidied up a bit...

Space before the `Config`.

Move the `PipelineConfig` description down to above the relevant struct field.

Add a similar description for the `ArgoCDConfig`

**Comment by wtam2018** [comment]
I think this PR should be merged together PR to change kustomization resource referances and  apply script.   If we don't merge together, we will give a broken system.   People can't try/play with this baby.   Perhaps, the two other PRs can be merged to a branch to run end-to-end tests before merging (together) to bootstrap?

**Review by bigkevmcd — APPROVED** [review]

File: pkg/pipelines/bootstrap.go
Comment (wtam2018) [reply pkg/pipelines/bootstrap.go]: @gaganeggday @ishitasequeira   Any thoughts?

**Review by wtam2018 — APPROVED** [review]

**Review by wtam2018 — CHANGES_REQUESTED** [review]
Don't merge until the following is ready.
* Dry-run script
* Set Kustomization references.

**Review by wtam2018 — APPROVED** [review]

**Pull request merged**
